lib/ex_unit/lib/ex_unit/on_exit_handler.ex

```diff
@@ -1,39 +1,75 @@
 defmodule ExUnit.OnExitHandler do
   @moduledoc false
+
   @name __MODULE__
+  @supervisor 2
+  @on_exit 3
 
   use Agent
 
+  @spec start_link(keyword()) :: {:ok, pid}
   def start_link(_opts) do
-    Agent.start_link(fn -> %{} end, name: @name)
+    Agent.start_link(fn ->
+      :ets.new(@name, [:public, :named_table, read_concurrency: true, write_concurrency: true])
+    end, name: @name)
   end
 
   @spec register(pid) :: :ok
   def register(pid) when is_pid(pid) do
-    Agent.update(@name, &Map.put(&1, pid, []))
+    :ets.insert(@name, {pid, nil, []})
   end
 
   @spec add(pid, term, (() -> term)) :: :ok | :error
   def add(pid, name_or_ref, callback) when is_pid(pid) and is_function(callback, 0) do
-    Agent.get_and_update(@name, fn map ->
-      if entries = Map.get(map, pid) do
+    try do
+      :ets.lookup_element(@name, pid, @on_exit)
+    rescue
+      _ -> :error
+    else
+      entries ->
         entries = List.keystore(entries, name_or_ref, 0, {name_or_ref, callback})
-        {:ok, Map.put(map, pid, entries)}
-      else
-        {:error, map}
-      end
-    end)
+        true = :ets.update_element(@name, pid, {@on_exit, entries})
+        :ok
```

Comment:
In the same way the ets API may have races and you guarantee it is safe on the usage.